### ISSUE TITLE
react/prop-types false positive with not exported functions

### ISSUE THREAD

**Issue by AuthorProxy** [yannickcr/eslint-plugin-react#1291]
```jsx
import axios from 'axios';
import React from 'react';

import TreeView from './TreeView';
import './TreeView.css';

const mapToNodes = entity => ({
  id: entity.id,
  name: entity.regionName,
  nodeState: entity.isExpanded ? 'expanded' : 'collapsed',
  children: Array.isArray(entity.children) ? entity.children.map(mapToNodes) : null,
  hasChildren: entity.hasChildren,
  hasChildrenValues: entity.hasChildrenValues
});

const getInnerNode = ({ node }) => node.name;
const getOuterNode = ({ nodeRow, nodeKey }) => (
  <div key={`${nodeKey}-row`} className="row tree-node__row">
    <div className="col-xs-6 tree-node__column">
      {nodeRow}
    </div>
    <div className="col-xs-3 tree-node__column">
      <div className="node">{`Column ${JSON.stringify(nodeKey)}`}</div>
    </div>
    <div className="col-xs-3 tree-node__column">
      <div className="node">{`Column ${JSON.stringify(nodeKey)}`}</div>
    </div>
  </div>
);

const getAsyncNodes = (parentNode) => {
  let promise;
  if (parentNode) {
    promise = axios.get(`/data/region-tags/region-tags-${parentNode.id}.json`);
  } else {
    promise = axios.get('/data/region-tags/region-tags-root.json');
  }

  return promise
    .then(response => response.data.map(mapToNodes))
    .catch((error) => {
      window.console.error(error);
      throw error;
    });
};

const TreeViewPage = () => (
  <div className="region-tag-tree">
    <h1 className="region-tag-tree__title">Региональные значения тегов</h1>
    <TreeView
      getNodes={getAsyncNodes}
      getInnerRender={getInnerNode}
      getOuterRender={getOuterNode}
    />
  </div>
);

export default TreeViewPage;
```

Given this code I got errors
```
  18:25  warning  'nodeRow' is missing in props validation  react/prop-types
  18:34  warning  'nodeKey' is missing in props validation  react/prop-types
```

at this string `const getOuterNode = ({ nodeRow, nodeKey }) => (` and not at others! It shouldn't verify not exported classes, because validation doesn't called by `React` (may be a bug too).

**Comment by ljharb** [comment]
That it's not exported is irrelevant; eslint-plugin-react identifies all jsx-returning functions as components - and there are plenty of components internal to a file that still need propTypes.

Why does TreeView take a function, instead of a React element?

**Comment by AuthorProxy** [comment]
1. It's pure function, that return react element, because another eslint rule said that you shouldn't use classes for react if there is no state.

2. It doesn't ask me to add props to `getInnerNode` that return raw string, that will be renderered the same as `getOuterNode` .

3. Even if I'll give prop-types they doesn't checked by react, I try it, so it useless

**Comment by ljharb** [comment]
A pure function that returns a React element is an SFC (Stateless Functional Component). SFCs have propTypes and defaultProps.

It'll be checked by React if `TreeView` renders it in a component context - if it doesn't, then it should.
